sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1007,6 +1007,8 @@ object CombineUnions extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transformDown {
     case u: Union => flattenUnion(u, false)
     case Distinct(u: Union) => Distinct(flattenUnion(u, true))
+    case Deduplicate(keys: Seq[Attribute], u: Union) if keys == u.output =>
```

Comment:
The current approach (@cloud-fan suggested in his comment) looks fine, but does this condition work correctly? What if they have a different order, `u.output` has new `exprId`s, ...?

Reply:
If `u.output`  has new expr IDs, then this plan will be unresolved...

Reply:
> If u.output has new expr IDs, then this plan will be unresolved...

Oh, I see. I missed that.

Reply:
I have added a test case for the different order case, the condition work fine, please review, @maropu @cloud-fan , thanks.